File: my-folder/0039-combination-sum/solution.java
class Solution {
    List<List<Integer>> res;
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        res=new ArrayList<>();
        int l =candidates.length;
        Arrays.sort(candidates);
        solve(0,candidates,target, new ArrayList<Integer>());

        return res;
    }

    void solve(int index, int[] candidates, int target, List<Integer> list){
         if(target==0){
            res.add(new ArrayList<>(list));
            return ;
        }
        else{
            for(int i=index;i<candidates.length;i++){
                if(candidates[i]>target)
                    break;
                list.add(candidates[i]);
                solve(i,candidates,target-candidates[i],list);
                list.remove(list.size()-1);
            }
        }

    }
}

File: my-folder/0039-combination-sum/solution.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        
        vector<vector<int>> res;
        vector<int>combination;
        sort(candidates.begin(),candidates.end());
        helper(candidates,target,res, combination, 0);
        return res;
        
    }
    
    void helper(vector<int>& candidates, int target, 
                       vector<vector<int>>& res,vector<int>& combination, int index)
    {
        if(target<=0)
        {
            res.push_back(combination);
            return;
        }
                      
        for(int i=index;i<candidates.size()&&target>=candidates[i];i++)
        {
            combination.push_back(candidates[i]);
            helper(candidates,target-candidates[i],res,combination, i);
            combination.pop_back();
        }
        
    }
};
